Add writing u.trajectory.ts.data['molecule_tag'] as molecule_tag atom attribute to LAMMPS datafile

mglagolev opened the pull request:
Fixes #3548

Being able to write the molecule id in a LAMMPS data file is crucial for using the MDAnalysis as a generator of LAMMPS initial configurations, for example, in machine learning workflows.
Currently, the molecule_tag field reads as atoms.resids, but when writing a data file, zero values are hardcoded.
Following the discussion in #3548, I've implemented writing ts.data['molecule_tag'] into this field. 

Changes made in this Pull Request:
 
Added data as input parameter for DATAWriter._write_atoms
If data['molecule_tag'] is present, _write_atoms will write its values as the molecule_tag attribute of the Atoms section of the data file. Otherwise, it will fill molecule_tag with zeroes, consistent with the previous behavior.

Added special read-write-read function to the tests, which sets data['molecule_tag'] values from atoms.resids after reading the data file. Added the class to check the resids after using the aforementioned function.

PR Checklist
------------
 - [V] Tests?
 - [X] Docs?
 - [V] CHANGELOG updated?
 - [V] Issue raised/referenced?


<!-- readthedocs-preview readthedocs-preview start -->
----
:books: Documentation preview :books:: https://readthedocs-preview--4114.org.readthedocs.build/en/4114/

<!-- readthedocs-preview readthedocs-preview end -->

mglagolev committed:
Add writing data['molecule_tag'] to LAMMPS datafile

orbeckst committed:
Merge branch 'develop' into lammps-write-molecule-tag

orbeckst commented:
@hmacdope do you have expertise with LAMMPS to review this PR, or can you suggest someone else?

RMeli commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -288,20 +288,27 @@ def _write_atoms(self, atoms):
-                             i=index, t=atype, c=charge, x=coords[0],
+            for index, moltag, atype, charge, coords in zip(indices, moltags,
+                    types, charges, coordinates):
+                self.f.write('{i:d} {m:d} {t:d} {c:f} {x:f} {y:f} {z:f}\n'.format(
+                             i=index, m=moltag, t=atype, c=charge, x=coords[0],
```
I'm wondering if with this many fields, switching to f-strings would be cleaner?

```python
'{i:d} {m:d} {t:d} {c:f} {x:f} {y:f} {z:f}\n'.format(i=index, m=moltag, t=atype, c=charge, x=coords[0], y=coords[1], z=coords[2])
```
vs
```python
x, y, z = coords
f"{index:d} {moltag:d} {atype:d} {charge:f} {x:f} {y:f} {z:f}\n"

RMeli commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -288,20 +288,27 @@ def _write_atoms(self, atoms):
         indices = atoms.indices + 1
         types = atoms.types.astype(np.int32)
 
+        try:
+            moltags = data['molecule_tag']
+        except KeyError:
+            moltags = [0] * len(atoms)
+
```
```suggestion
        moltags = data.get("molecule_tag", np.zeros(len(atoms), dtype=int))
```

Allocating a `numpy` array should be considerably faster. `get` allows to avoid the try-except logic.

```
$ N = 10000
$ %timeit [0] * N
23.5 µs ± 359 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

$ %timeit np.zeros(N, dtype=int)
2.85 µs ± 231 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

mglagolev replied:
Thanks for this suggestion! This is clearly a better way.

RMeli commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -397,6 +404,8 @@ def write(self, selection, frame=None):
         else:
             frame = u.trajectory.ts.frame
 
+        data = u.trajectory.ts.data
```
Is this necessary since it `data` is only used one time?

RMeli commented on testsuite/MDAnalysisTests/coordinates/test_lammps.py:
```diff
@@ -115,6 +115,29 @@ def LAMMPSDATAWriter(request, tmpdir_factory):
 
 
+@pytest.fixture(params=[
+    LAMMPSdata,
+    LAMMPSdata_mini,
+    LAMMPScnt,
+    LAMMPShyd,
```
From the code coverage comment above, it seems that none of those covers the case where there is no charge. Can we add one such case?

RMeli reviewed (commented):
A few small comments. I'll leave it to LAMMPS experts for a more in-depth review.

mglagolev committed:
Update package/MDAnalysis/coordinates/LAMMPS.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mglagolev commented:
@hmacdope  Shall I improve something in this PR?

hmacdope reviewed (changes requested):
@mglagolev sorry for the delay, looks good, if you could address @RMeli's last comments about f-strings and a test case coverage for a case with no carge we should be good to go

mglagolev committed:
Improving code based on RMeli's suggestions

mglagolev commented:
Thank you @hmacdope and, once again, @RMeli!
I've added the @RMeli's improvements, except running a test on an uncharged system.
There is an uncharged datafile, which I added with my previous commit, but it is compressed.
When adding it to the test fixture, I run into an error, apparently because the filename upon writing is changed from .data.bz2 to .data.data

Here's a simple script to test this behavior:

```
import MDAnalysis as mda
import numpy as np
u = mda.Universe.empty(1, trajectory = True)
u.add_TopologyAttr('type')
u.add_TopologyAttr('mass')
atom = mda.core.groups.Atom(u = u, ix = 0)
atom.position = np.array([0., 0., 0.])
u.atoms = mda.AtomGroup([atom,])
u.add_bonds([])
u.add_angles([])
u.add_dihedrals([])
u.add_impropers([])
u.atoms.types = ['1',]
u.atoms.masses = [1.,]
u.dimensions = [1, 1, 1, 90, 90, 90]
with mda.Writer("test.data.bz2", n_atoms = 1) as w:
	w.write(u.atoms)
```
All the attributes are set just to get the simplest datafile written without an error.
In my environments, using both the development branch and the version from pip, the datafiles are written under the name "test.data.data"
Am I missing something? Is it a bug or a feature?

mglagolev commented:
Maybe @fenilsuchak can help?

RMeli commented:
Sounds odd, I'll have to double check this behaviour. Thanks for checking and reporting.

A possible workaround/alternative to get this PR done is to use `del_TopologyAttr` on the universe obtained from one of the current files to remove the `charge` attribute.

```python
In [1]: from MDAnalysisTests.datafiles import LAMMPSdata

In [2]: import MDAnalysis as mda

In [3]: u = mda.Universe(LAMMPSdata)

In [4]: u.atoms.charges
Out[4]: array([0., 0., 0., ..., 0., 0., 0.])

In [5]: u.del_TopologyAttr('charges')

In [6]: u.atoms.charges
---------------------------------------------------------------------------
NoDataError                               Traceback (most recent call last)
Cell In[6], line 1
----> 1 u.atoms.charges

File ~/Documents/git/mdanalysis/package/MDAnalysis/core/groups.py:2539, in AtomGroup.__getattr__(self, attr)
   2537 elif attr == 'positions':
   2538     raise NoDataError('This Universe has no coordinates')
-> 2539 return super(AtomGroup, self).__getattr__(attr)

File ~/Documents/git/mdanalysis/package/MDAnalysis/core/groups.py:614, in GroupBase.__getattr__(self, attr)
    612     else:
    613         err = 'This Universe does not contain {singular} information'
--> 614         raise NoDataError(err.format(singular=cls.singular))
    615 else:
    616     return super(GroupBase, self).__getattr__(attr)

NoDataError: This Universe does not contain charge information
```

mglagolev committed:
Adding a test on a system with deleted charges

mglagolev commented:
@RMeli, thanks for the idea!
The test is passed without an error.

RMeli commented on testsuite/MDAnalysisTests/coordinates/test_lammps.py:
```diff
@@ -115,6 +115,32 @@ def LAMMPSDATAWriter(request, tmpdir_factory):
+def LAMMPSDATAWriter_molecule_tag(request, tmpdir_factory):
+    filename, charges = request.param
+    u = mda.Universe(filename)
+    if charges == False:
```
```suggestion
    if not charges:
```

IAlibay replied:
Please review https://github.com/MDAnalysis/mdanalysis/actions/runs/4859346505/jobs/8661858282 this was highlighted by the darker linting bot as a flake8 failure (see: https://github.com/MDAnalysis/mdanalysis/pull/4114#issuecomment-1496674113)

mglagolev replied:
Oh, thanks again to @RMeli. Flake8 should pass fine now.

RMeli reviewed (approved):
Thanks @mglagolev. Small nitpick (you can safely ignore it), otherwise LGTM.

mglagolev committed:
Update testsuite/MDAnalysisTests/coordinates/test_lammps.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mglagolev commented:
@IAlibay, sorry for bothering, is anything else required to merge this?

RMeli commented:
Thanks for the contribution @mglagolev, and sorry if it took quite long to get it merged.

mglagolev commented:
Great! Thanks to helpful people here for making this small contribution possible:-)

By the way, @RMeli  did you reproduce the issue with compressed filenames?

RMeli commented:
@mglagolev unfortunately no, I forgot. Thanks for reminding me! If you don't mind, could you please open an issue with your observation? Otherwise I'll open one.

mglagolev commented:
@rmeli Sure!
https://github.com/MDAnalysis/mdanalysis/issues/4159
